### PR TITLE
Avoid ambiguous references in proxy classes

### PR DESCRIPTION
**Pull request by kneth**
#686 reports ambiguous references in proxy classes. To avoid this, only import required classes.

@emanuelez @cmelchior @bmunkholm

### CODE REVIEW

**Comment by kneth** [comment]
@cmelchior @emanuelez @bmunkholm Ready for review.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 1]
```diff
@@ -155,22 +156,42 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+
+        ArrayList<String> imports = new ArrayList<String>();
+        imports.add("android.util.JsonReader");
+        imports.add("android.util.JsonToken");
+        imports.add("io.realm.RealmObject");
+        imports.add("io.realm.internal.ColumnType");
+        imports.add("io.realm.internal.Table");
+        imports.add("io.realm.internal.ImplicitTransaction");
+        imports.add("io.realm.internal.LinkView");
+        imports.add("io.realm.internal.android.JsonUtils");
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: io.realm.internal should be rearrange so it is in alphabetical order. IDE's does this automatically and it is a pain when editing the unit tests if they are not in alfabetical order.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 2]
```diff
@@ -155,22 +156,42 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+        imports.add("java.io.IOException");
+        imports.add("java.util.*");
+        imports.add("org.json.JSONObject");
+        imports.add("org.json.JSONException");
+        imports.add("org.json.JSONArray");
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: same for org.json

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 3]
```diff
@@ -20,6 +20,7 @@
 
 import java.io.BufferedWriter;
 import java.io.IOException;
+import java.lang.String;
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: No need for this. `java.lang.*` is imported by default.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 4]
```diff
@@ -155,22 +156,42 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+            if (typeUtils.isAssignable(field.asType(), realmObject))  { // Links
+                fieldTypeName = field.asType().toString();
+                if (!imports.contains(fieldTypeName)) {
+                    imports.add(fieldTypeName);
+                }
+            } else if (typeUtils.isAssignable(field.asType(), realmList)) { // LinkLists
+                String fullType = field.asType().toString();
+                int index = fullType.indexOf('<');
+                int jndex = fullType.indexOf('>');
+                fieldTypeName = fullType.substring(index + 1, jndex);
```
Comment (emanuelez) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Can't you use `getGenericType()` ?

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 5]
```diff
@@ -155,22 +158,42 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+            if (typeUtils.isAssignable(field.asType(), realmObject)) { // Links
+                fieldTypeName = field.asType().toString();
+            } else if (typeUtils.isAssignable(field.asType(), realmList)) { // LinkLists
+                fieldTypeName = ((DeclaredType) field.asType()).getTypeArguments().get(0).toString();
+            }
+            if (fieldTypeName != "" && !imports.contains(fieldTypeName)) {
+                imports.add(fieldTypeName);
+            }
+        }
+        imports.sort(new Comparator<String>() {
```
Comment (cmelchior) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: You can use Collections.sort(imports) as a one-liner.

File: realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java [hunk 6]
```diff
@@ -155,22 +158,42 @@ public void generate() throws IOException, UnsupportedOperationException {
...
+        imports.add("android.util.JsonReader");
+        imports.add("android.util.JsonToken");
+        imports.add("io.realm.RealmObject");
+        imports.add("io.realm.internal.ColumnType");
+        imports.add("io.realm.internal.Table");
+        imports.add("io.realm.internal.ImplicitTransaction");
+        imports.add("io.realm.internal.LinkView");
+        imports.add("io.realm.internal.android.JsonUtils");
+        imports.add("java.io.IOException");
+        imports.add("java.util.*");
```
Comment (bmunkholm) [inline realm-annotations-processor/src/main/java/io/realm/processor/RealmProxyClassGenerator.java]: Why include util.* and not be specific?

**Pull request merged**